### ISSUE TITLE
Feature request: Bayesian BM25

### ISSUE THREAD

**Issue by ccdv-ai** [neuml/txtai#1023]
Add [Bayesian BM25](https://www.researchgate.net/publication/400212695_Bayesian_BM25_A_Probabilistic_Framework_for_Hybrid_Text_and_Vector_Search), a bounded score ([0,1]) that transforms BM25 relevance scores into calibrated probability estimates. Can be useful to merge scores for hybrid search.

Implementation is inspired by this [code](https://github.com/instructkr/bb25/blob/main/src/bayesian_scorer.rs), should be something like: 
```python
import numpy as np

from .bm25 import BM25

class BayesianBM25(BM25):
    """
    Bayesian BM25 scoring.
    This class converts the standard BM25 score into a posterior probability.
    """

    def __init__(self, config=None):
        super().__init__(config)
        
        # Bayesian parameters:
        # alpha: controls the steepness of the sigmoid curve
        # beta: the raw BM25 score value that results in a 0.5 probability
        self.alpha = self.config.get("alpha", 1.0)
        self.beta = self.config.get("beta", 1.0)

    def score(self, freq, idf, length):
        """
        Calculates the Bayesian Posterior Probability for a single term.
        Matches the signature of the base BM25 class.
        """
        # 1. Calculate the standard BM25 score using the parent class method
        raw_score = super().score(freq, idf, length)
        
        if raw_score <= 0:
            return 0.0

        # 2. Likelihood: Transform raw score to probability using a Sigmoid function
        # P(Score | Relevant)
        likelihood = 1 / (1 + np.exp(-self.alpha * (raw_score - self.beta)))

        # 3. Prior: Estimate P(Relevant) based on doc structure (TF and Length)
        # Prior based on term frequency
        p_tf = 0.2 + 0.7 * min(freq / 10.0, 1.0)
        
        # Prior based on length normalization
        avgdl = self.avgdl if self.avgdl > 0 else 1.0
        ratio = length / avgdl
        p_norm = np.clip(1.0 / (1.0 + ratio), 0.1, 0.9)
        
        # Composite Prior
        prior = np.clip(0.7 * p_tf + 0.3 * p_norm, 0.1, 0.9)

        # 4. Posterior: Apply Bayes' Theorem
        # P(Relevant | Score) = (Likelihood * Prior) / Evidence
        numerator = likelihood * prior
        denominator = numerator + (1.0 - likelihood) * (1.0 - prior)
        
        return numerator / (denominator + 1e-10)
```

There is also an additional method to merge scores https://github.com/instructkr/bb25/pull/1

**Comment by davidmezzetti** [comment]
Excellent. Then this change could be a new normalization method and benefit all term searches! 

This change can then work on this section: https://github.com/neuml/txtai/blob/master/src/python/txtai/scoring/tfidf.py#L169

I could see a new normalize class in the scoring folder being a good way to encapsulate this logic.

**Comment by jaepil** [comment]
@davidmezzetti

Here's my conclusion. I will open the PR tomorrow.

## 1. Root Cause: Fusion Mismatch

BB25's normalize step outputs **calibrated probabilities** in [0, 1]. But txtai's hybrid search fuses dense and sparse scores via **convex combination**:

    score = w * dense + (1 - w) * sparse

Convex combination treats scores as raw weights. It works well when both signals are on comparable scales (as with the default min-max normalization), but it discards the Bayesian probability semantics that BB25 provides.

The reference implementation (cognica-io/bayesian-bm25) uses **log-odds conjunction** from Paper 2 ("From Bayesian Inference to Neural Computation"):

    logit(p) = log(p / (1 - p))
    l_bar    = w_d * logit(p_dense) + w_s * logit(p_sparse)
    l_adj    = l_bar * n^alpha          # confidence scaling, n=2 signals, alpha=0.5

This fuses evidence in logit space where independent probability signals combine additively -- the mathematically correct way to accumulate Bayesian evidence.

## 2. The Fix: Dual Dynamic Calibration + Log-Odds

### Problem with naive log-odds

A naive implementation using `(1 + cos) / 2` to map dense cosine scores to probabilities **made things worse**. Distribution analysis revealed why:

- Dense cosine scores are always > 0 for retrieved candidates
- `(1 + cos) / 2` with cos > 0 always gives p > 0.5, meaning **100% positive logits**
- BB25 sparse logits are centered at 0 (50% positive, 50% negative)
- Dense signal completely **dominates** in logit space

### Solution: symmetric calibration

Apply the same per-query dynamic sigmoid calibration to dense cosine scores that BB25 applies to BM25 scores:

    logit_dense = alpha_eff * (cos_score - median_score)

where `alpha_eff = 1 / std(scores)` and `median_score = median(positive scores)` for the current query's candidate set. This centers the dense logit at 0 for the median candidate, making it symmetric with the sparse BB25 logit.

### Raw logit ranking

The fused score is returned as `l_adj` directly (raw logit) rather than `sigmoid(l_adj)`. Applying sigmoid would compress top candidates into a narrow range near 1.0, destroying fine-grained ranking resolution. Since the fused score is only used for ranking (not as a calibrated probability), the raw logit preserves all ordering information.

## 3. Final Benchmark Results (BEIR, nDCG@10)

| Dataset  | Default+Convex | BB25+Convex | BB25+LogOdds | vs Default |
|----------|:-:|:-:|:-:|:-:|
| arguana  | 49.47%         | 51.08%      | **51.70%**   | **+2.23**  |
| fiqa     | 35.41%         | 32.91%      | **37.44%**   | **+2.03**  |
| nfcorpus | **34.79%**     | 34.14%      | 32.83%       | -1.96      |
| scidocs  | 19.71%         | 18.71%      | **20.33%**   | **+0.62**  |
| scifact  | 71.08%         | 69.70%      | **72.41%**   | **+1.33**  |

**4/5 datasets improved over Default.** Average delta: +0.85 across all 5 datasets.

BB25+LogOdds vs BB25+Convex recovery:
- fiqa: 32.91 -> 37.44 (+4.53, fully recovered and surpassed)
- scidocs: 18.71 -> 20.33 (+1.62, fully recovered and surpassed)
- scifact: 69.70 -> 72.41 (+2.71, fully recovered and surpassed)
- arguana: 51.08 -> 51.70 (+0.62, further improved)

## 4. nfcorpus Regression Analysis

nfcorpus is the only dataset that regresses (-1.96 vs Default). Analysis shows this is **not a calibration or saturation problem** but a structural mismatch between log-odds fusion and nfcorpus's unique characteristics:

### Dataset characteristics

| Property                          | nfcorpus   | scifact (comparison) |
|-----------------------------------|------------|----------------------|
| Corpus size                       | 3,633      | 5,183                |
| Query length (median words)       | **2**      | 12                   |
| Relevant docs per query (mean)    | **38.2**   | 1.1                  |
| Relevance levels                  | Graded (1,2) | Binary             |

### Why log-odds hurts nfcorpus

1. **Many relevant docs per query (38.2):** Top-10 results are mostly all relevant. NDCG depends on ordering *among* them, not on retrieving them. Log-odds changes the relative ordering in logit space compared to convex combination.

2. **Graded relevance (levels 1 and 2):** Placing grade-2 (highly relevant) above grade-1 matters for NDCG. The nonlinear logit transform can reorder documents whose scores differ only slightly in probability space.

3. **Very short queries (median 2 words):** Queries like "Breast Cancer" produce many medical documents with near-identical BM25 scores. Small score differences get amplified nonlinearly by the logit transform.

### Attempted mitigations (no effect)

- **Logit cap (+-6):** Individual logits already fall within this range due to the dynamic calibration. No change in nfcorpus scores.
- **Tanh soft clamping:** `tau * tanh(l_adj / tau)` produces identical results to raw logit for the logit magnitudes observed.

### Conclusion

The nfcorpus regression is an inherent property of logit-space fusion on corpora with very short queries, many relevant documents, and graded relevance. The nonlinear logit transform reorders documents whose scores are very close, and this reordering happens to be slightly worse for nfcorpus's graded NDCG evaluation. This is a tradeoff: the same nonlinearity that improves 4/5 datasets (by amplifying signal agreement) slightly hurts the one dataset where fine-grained ordering among many high-scoring relevant documents matters most.

## 5. Evolution of Approaches

| Version          | arguana | fiqa   | nfcorpus | scidocs | scifact | Wins |
|------------------|---------|--------|----------|---------|---------|------|
| Default+Convex   | 49.47   | 35.41  | **34.79**| 19.71   | 71.08   | --   |
| BB25+Convex      | 51.08   | 32.91  | 34.14    | 18.71   | 69.70   | 1/5  |
| BB25+Sigmoid     | 51.51   | 35.54  | 33.68    | 20.06   | 71.43   | 4/5  |
| BB25+Raw Logit   |**51.70**|**37.44**| 32.83   |**20.33**|**72.41**| 4/5  |
| BB25+Logit Cap   | 51.70   | 37.46  | 32.83    | 20.33   | 72.65   | 4/5  |
| BB25+Tanh Clamp  | 51.70   | 37.44  | 32.83    | 20.33   | 72.41   | 4/5  |

Raw Logit was selected as the final version: simplest code, best overall performance, no arbitrary hyperparameters (cap/tau). It can be easily converted to probability by using `sigmoid(logit)`.
